renderdoc/driver/ihv/nv/nv_counters.cpp

```diff
@@ -0,0 +1,380 @@
+#include "nv_counters.h"
+#include "common/common.h"
+
+#define NVPM_INITGUID
+#include "official/PerfKit/include/NvPmApi.h"
+#include "strings/string_utils.h"
+
+#include <algorithm>
+
+struct EnumCountersCtx
+{
+  std::vector<GPUCounter> mExternalIds;
+  std::vector<uint32_t> mInternalIds;
+  std::vector<CounterDescription> mExternalDescriptors;
+  std::vector<uint32_t> mInternalDescriptors;
+
+  NvPmApi *mNvPmApi;
+
+  uint32_t mCurrentCounterId;
+};
+
+static bool NvPmResultFails(NVPMRESULT actual, char const *failMsg)
+{
+  if(actual != NVPM_OK)
+  {
+    RDCWARN("NV GPU performance counters could not %s (code = %u)\n", failMsg, actual);
+    return true;
+  }
+  return false;
+}
+
+int NvPmCountCounters(NVPMCounterID unCounterID, const char *pcCounterName, void *pUserData)
+{
+  uint32_t *pNumCounters = static_cast<uint32_t *>(pUserData);
```

Comment:
Please prefer normal casts `= (uint32_t *)pUserData;` over C++ style casts. There are a few other instances in this file.